File: trilogy/io/adapters.py
"""Turn whatever a source function returned into an Arrow batch reader.

Probes are duck-typed and ordered, so no optional dependency is imported unless
an object of that kind actually shows up. Streaming inputs stay streaming --
nothing here materializes a generator into a table.

A declared schema wins over whatever the source produced, whichever adapter
handled it: ``to_reader(obj, schema).schema`` is always ``schema``.
"""

from __future__ import annotations

from collections.abc import Callable, Iterator, Mapping
from typing import Any

import pyarrow as pa

from trilogy.io.errors import TrilogyIOError

Probe = Callable[[Any], bool]
Convert = Callable[[Any, "pa.Schema | None"], pa.RecordBatchReader]

_ADAPTERS: list[tuple[Probe, Convert]] = []


def register_adapter(predicate: Probe, converter: Convert | None = None) -> Any:
    """Teach the wrapper about another return type.

    Usable directly or as a decorator. Later registrations are probed first, so
    a caller can override a built-in.
    """
    if converter is not None:
        _ADAPTERS.append((predicate, converter))
        return converter

    def decorate(fn: Convert) -> Convert:
        _ADAPTERS.append((predicate, fn))
        return fn

    return decorate


def to_reader(obj: Any, schema: pa.Schema | None = None) -> pa.RecordBatchReader:
    for predicate, converter in reversed(_ADAPTERS):
        if predicate(obj):
            return _conform_reader(converter(obj, schema), schema)
    raise TrilogyIOError(
        f"Cannot convert {type(obj).__module__}.{type(obj).__qualname__} to an "
        "Arrow stream. Return a pyarrow Table, a dataframe implementing the "
        "Arrow PyCapsule interface, a list of dicts, or an iterator of any of "
        "those -- or register_adapter() your own."
    )


def empty_reader(schema: pa.Schema | None) -> pa.RecordBatchReader:
    if schema is None:
        raise TrilogyIOError(
            "Cannot infer a schema from an empty result. Pass schema= so the "
            "source still describes its columns when it has no rows."
        )
    return pa.RecordBatchReader.from_batches(schema, iter(()))


def conform(batch: pa.RecordBatch, schema: pa.Schema) -> pa.RecordBatch:
    """Cast a batch onto the stream's schema; a no-op in the common case.

    Columns are cast in place rather than through a Table: a 0-row Table yields
    no batches at all, so the round trip loses the empty batch entirely.
    """
    if batch.schema.equals(schema):
        return batch
    if batch.schema.names != schema.names:
        batch = batch.select(schema.names)
    return pa.RecordBatch.from_arrays(
        [column.cast(field.type) for column, field in zip(batch.columns, schema)],
        schema=schema,
    )


def _conform_reader(
    reader: pa.RecordBatchReader, schema: pa.Schema | None
) -> pa.RecordBatchReader:
    """Make a declared schema authoritative over whatever the source produced.

    Applied here rather than in each adapter, because half of them carry a
    schema of their own and would otherwise silently win: a source returning a
    pyarrow table and one returning dicts must answer ``--describe`` the same
    way they fill the stream. The declared schema also fixes column order, so
    it is a projection as much as a cast.
    """
    if schema is None or reader.schema.equals(schema):
        return reader
    missing = [name for name in schema.names if name not in reader.schema.names]
    if missing:
        raise TrilogyIOError(
            f"Source produced columns {reader.schema.names}, which do not cover "
            f"its declared schema: {', '.join(missing)} missing."
        )
    return pa.RecordBatchReader.from_batches(
        schema, (conform(batch, schema) for batch in reader)
    )


def _module_root(obj: Any) -> str:
    return type(obj).__module__.partition(".")[0]


def _chain(
    head: pa.RecordBatchReader, rest: Iterator[Any], schema: pa.Schema
) -> Iterator[pa.RecordBatch]:
    for batch in head:
        yield conform(batch, schema)
    for item in rest:
        for batch in to_reader(item, schema):
            yield conform(batch, schema)


def _from_iterable(obj: Any, schema: pa.Schema | None) -> pa.RecordBatchReader:
    """Peek one element to learn the schema, then stream the rest behind it."""
    iterator = iter(obj)
    try:
        first = next(iterator)
    except StopIteration:
        return empty_reader(schema)
    head = to_reader(first, schema)
    resolved = schema or head.schema
    return pa.RecordBatchReader.from_batches(resolved, _chain(head, iterator, resolved))


def _from_pylist(obj: Any, schema: pa.Schema | None) -> pa.RecordBatchReader:
    if not obj:
        return empty_reader(schema)
    return pa.Table.from_pylist(list(obj), schema=schema).to_reader()


def _from_pydict(obj: Any, schema: pa.Schema | None) -> pa.RecordBatchReader:
    return pa.Table.from_pydict(dict(obj), schema=schema).to_reader()


def _from_capsule_array(obj: Any, schema: pa.Schema | None) -> pa.RecordBatchReader:
    batch = pa.record_batch(obj)
    return pa.RecordBatchReader.from_batches(batch.schema, iter((batch,)))


def _from_pandas(obj: Any, schema: pa.Schema | None) -> pa.RecordBatchReader:
    return pa.Table.from_pandas(obj, schema=schema, preserve_index=False).to_reader()


# --- built-in probes ---------------------------------------------------------
# Registration is most-generic first, because probing runs in reverse: the last
# registration wins, which is what lets a caller override a built-in. Keep the
# capsule probes below the concrete pyarrow ones -- a pa.Table exposes
# __arrow_c_stream__ too, but to_reader() preserves its existing chunking.

register_adapter(
    lambda o: isinstance(o, (list, tuple)) or hasattr(o, "__next__"), _from_iterable
)
register_adapter(lambda o: isinstance(o, Mapping), _from_pydict)
register_adapter(
    lambda o: isinstance(o, (list, tuple))
    and all(isinstance(item, Mapping) for item in o),
    _from_pylist,
)
register_adapter(lambda o: hasattr(o, "__arrow_c_array__"), _from_capsule_array)
register_adapter(
    lambda o: hasattr(o, "__arrow_c_stream__"),
    lambda o, schema: pa.RecordBatchReader.from_stream(o),
)
# pandas 2 predates the capsule interface. Matched on the module name so pandas
# is never imported on behalf of a script that does not use it.
register_adapter(
    lambda o: _module_root(o) == "pandas" and type(o).__name__ == "DataFrame",
    _from_pandas,
)
register_adapter(
    lambda o: isinstance(o, pa.RecordBatch),
    lambda o, schema: pa.RecordBatchReader.from_batches(o.schema, iter((o,))),
)
register_adapter(lambda o: isinstance(o, pa.Table), lambda o, schema: o.to_reader())
register_adapter(lambda o: isinstance(o, pa.RecordBatchReader), lambda o, schema: o)
register_adapter(lambda o: o is None, lambda o, schema: empty_reader(schema))

File: crates/trilogy-io/src/adapters.rs
//! Turn whatever a source returned into an Arrow batch reader.
//!
//! The Rust analogue of the python adapter registry: a trait bound rather than
//! a probe list, since the type is known at compile time. Streaming inputs stay
//! streaming -- nothing here collects a reader into memory.
//!
//! A declared schema wins over whatever the source produced, whichever impl
//! handled it -- see [`conform_reader`], which `resolve` applies centrally.

use arrow::array::{RecordBatch, RecordBatchIterator, RecordBatchReader};
use arrow::datatypes::SchemaRef;
use arrow::error::ArrowError;
use arrow::ffi_stream::{ArrowArrayStreamReader, FFI_ArrowArrayStream};

use crate::error::{Error, Result};

pub type BatchReader = Box<dyn RecordBatchReader + Send>;

pub trait IntoBatchReader {
    /// `schema` is the source's declared schema, if it has one. Impls only need
    /// it to describe an empty result; [`conform_reader`] enforces it on the
    /// rows themselves, so ignoring it here is safe.
    fn into_batch_reader(self, schema: Option<SchemaRef>) -> Result<BatchReader>;
}

/// Make a declared schema authoritative over whatever the source produced.
///
/// Applied at one seam rather than in each impl, because most of them carry a
/// schema of their own and would otherwise silently win: a source returning a
/// `RecordBatch` and one returning `Vec<RecordBatch>` must answer `--describe`
/// the same way they fill the stream. The declared schema also fixes column
/// order and drops what it does not name, so it is a projection as much as a
/// cast. Mirrors `trilogy.io.adapters.to_reader`.
pub fn conform_reader(reader: BatchReader, schema: Option<SchemaRef>) -> Result<BatchReader> {
    let Some(schema) = schema else {
        return Ok(reader);
    };
    let produced = reader.schema();
    if produced == schema {
        return Ok(reader);
    }
    let missing: Vec<&str> = schema
        .fields()
        .iter()
        .filter(|field| produced.field_with_name(field.name()).is_err())
        .map(|field| field.name().as_str())
        .collect();
    if !missing.is_empty() {
        let names: Vec<&str> = produced
            .fields()
            .iter()
            .map(|field| field.name().as_str())
            .collect();
        return Err(Error::adapter(format!(
            "Source produced columns [{}], which do not cover its declared \
             schema: {} missing.",
            names.join(", "),
            missing.join(", ")
        )));
    }
    Ok(Box::new(Conformed {
        inner: reader,
        schema,
    }))
}

fn conform(batch: RecordBatch, schema: &SchemaRef) -> std::result::Result<RecordBatch, ArrowError> {
    if batch.schema() == *schema {
        return Ok(batch);
    }
    let columns = schema
        .fields()
        .iter()
        .map(|field| {
            let column = batch.column_by_name(field.name()).ok_or_else(|| {
                ArrowError::InvalidArgumentError(format!(
                    "batch is missing declared column {}",
                    field.name()
                ))
            })?;
            arrow::compute::cast(column, field.data_type())
        })
        .collect::<std::result::Result<Vec<_>, ArrowError>>()?;
    // `try_new` rejects a 0-row batch built from no columns, so an empty schema
    // has to keep the row count it came in with.
    RecordBatch::try_new_with_options(
        schema.clone(),
        columns,
        &arrow::array::RecordBatchOptions::new().with_row_count(Some(batch.num_rows())),
    )
}

struct Conformed {
    inner: BatchReader,
    schema: SchemaRef,
}

impl Iterator for Conformed {
    type Item = std::result::Result<RecordBatch, ArrowError>;

    fn next(&mut self) -> Option<Self::Item> {
        let batch = self.inner.next()?;
        Some(batch.and_then(|b| conform(b, &self.schema)))
    }
}

impl RecordBatchReader for Conformed {
    fn schema(&self) -> SchemaRef {
        self.schema.clone()
    }
}

fn empty(schema: Option<SchemaRef>) -> Result<BatchReader> {
    let schema = schema.ok_or_else(|| {
        Error::adapter(
            "Cannot infer a schema from an empty result. Set .schema() so the \
             source still describes its columns when it has no rows.",
        )
    })?;
    Ok(Box::new(RecordBatchIterator::new(
        std::iter::empty(),
        schema,
    )))
}

impl IntoBatchReader for BatchReader {
    fn into_batch_reader(self, _schema: Option<SchemaRef>) -> Result<BatchReader> {
        Ok(self)
    }
}

impl IntoBatchReader for RecordBatch {
    fn into_batch_reader(self, _schema: Option<SchemaRef>) -> Result<BatchReader> {
        let schema = self.schema();
        Ok(Box::new(RecordBatchIterator::new(
            std::iter::once(Ok(self)),
            schema,
        )))
    }
}

impl IntoBatchReader for Vec<RecordBatch> {
    fn into_batch_reader(self, schema: Option<SchemaRef>) -> Result<BatchReader> {
        let Some(first) = self.first() else {
            return empty(schema);
        };
        let schema = schema.unwrap_or_else(|| first.schema());
        Ok(Box::new(RecordBatchIterator::new(
            self.into_iter().map(Ok),
            schema,
        )))
    }
}

impl IntoBatchReader for Option<RecordBatch> {
    fn into_batch_reader(self, schema: Option<SchemaRef>) -> Result<BatchReader> {
        match self {
            Some(batch) => batch.into_batch_reader(schema),
            None => empty(schema),
        }
    }
}

/// So a fallible source can be handed over directly, without unwrapping first.
impl<T, E> IntoBatchReader for std::result::Result<T, E>
where
    T: IntoBatchReader,
    E: std::error::Error + Send + Sync + 'static,
{
    fn into_batch_reader(self, schema: Option<SchemaRef>) -> Result<BatchReader> {
        match self {
            Ok(value) => value.into_batch_reader(schema),
            Err(e) => Err(Error::Source(Box::new(e))),
        }
    }
}

/// The Rust side of the same C stream interface python reads through
/// `__arrow_c_stream__` -- polars and datafusion both hand one over.
impl IntoBatchReader for FFI_ArrowArrayStream {
    fn into_batch_reader(self, _schema: Option<SchemaRef>) -> Result<BatchReader> {
        Ok(Box::new(ArrowArrayStreamReader::try_new(self)?))
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use arrow::array::{Int32Array, Int64Array, StringArray};
    use arrow::datatypes::{DataType, Field, Schema};

    use super::*;

    fn batch() -> RecordBatch {
        let schema = Arc::new(Schema::new(vec![Field::new("i", DataType::Int64, false)]));
        RecordBatch::try_new(schema, vec![Arc::new(Int64Array::from(vec![1, 2, 3]))]).unwrap()
    }

    fn count(reader: BatchReader) -> usize {
        reader.map(|b| b.unwrap().num_rows()).sum()
    }

    #[test]
    fn converts_the_supported_shapes() {
        assert_eq!(count(batch().into_batch_reader(None).unwrap()), 3);
        assert_eq!(
            count(vec![batch(), batch()].into_batch_reader(None).unwrap()),
            6
        );
        assert_eq!(count(Some(batch()).into_batch_reader(None).unwrap()), 3);
    }

    #[test]
    fn a_fallible_source_passes_through() {
        let ok: std::result::Result<RecordBatch, std::io::Error> = Ok(batch());
        assert_eq!(count(ok.into_batch_reader(None).unwrap()), 3);

        let failed: std::result::Result<RecordBatch, std::io::Error> =
            Err(std::io::Error::other("nope"));
        assert!(failed.into_batch_reader(None).is_err());
    }

    #[test]
    fn an_empty_result_needs_a_declared_schema() {
        assert!(Vec::<RecordBatch>::new().into_batch_reader(None).is_err());
        let schema = batch().schema();
        assert_eq!(
            count(
                Vec::<RecordBatch>::new()
                    .into_batch_reader(Some(schema))
                    .unwrap()
            ),
            0
        );
    }

    // --- a declared schema is authoritative ---------------------------------

    fn declared() -> SchemaRef {
        Arc::new(Schema::new(vec![
            Field::new("i", DataType::Int64, true),
            Field::new("s", DataType::Utf8, true),
        ]))
    }

    /// Deliberately the wrong types, wrong order, and one column too many.
    fn divergent() -> RecordBatch {
        let schema = Arc::new(Schema::new(vec![
            Field::new("s", DataType::Utf8, true),
            Field::new("extra", DataType::Int64, true),
            Field::new("i", DataType::Int32, true),
        ]));
        RecordBatch::try_new(
            schema,
            vec![
                Arc::new(StringArray::from(vec!["a", "b"])),
                Arc::new(Int64Array::from(vec![9, 9])),
                Arc::new(Int32Array::from(vec![1, 2])),
            ],
        )
        .unwrap()
    }

    fn conformed(batch: RecordBatch) -> BatchReader {
        let reader = batch.into_batch_reader(None).unwrap();
        conform_reader(reader, Some(declared())).unwrap()
    }

    #[test]
    fn a_declared_schema_casts_reorders_and_narrows() {
        let mut reader = conformed(divergent());
        assert_eq!(reader.schema(), declared());
        let batch = reader.next().unwrap().unwrap();
        assert_eq!(batch.schema(), declared());
        assert_eq!(
            batch
                .column(0)
                .as_any()
                .downcast_ref::<Int64Array>()
                .unwrap()
                .values(),
            &[1, 2]
        );
    }

    #[test]
    fn a_matching_schema_is_not_rewrapped() {
        let reader = batch().into_batch_reader(None).unwrap();
        let schema = reader.schema();
        assert_eq!(count(conform_reader(reader, Some(schema)).unwrap()), 3);
    }

    #[test]
    fn no_declared_schema_leaves_the_source_alone() {
        let reader = divergent().into_batch_reader(None).unwrap();
        let produced = reader.schema();
        assert_eq!(conform_reader(reader, None).unwrap().schema(), produced);
    }

    #[test]
    fn a_column_the_declared_schema_names_must_exist() {
        let reader = batch().into_batch_reader(None).unwrap();
        let Err(error) = conform_reader(reader, Some(declared())) else {
            panic!("a declared column the source omits must be an error");
        };
        assert!(error.to_string().contains("s missing"), "{error}");
    }

    #[test]
    fn conforming_preserves_an_empty_batch() {
        let source = Arc::new(Schema::new(vec![Field::new("i", DataType::Int32, true)]));
        let empty =
            RecordBatch::try_new(source, vec![Arc::new(Int32Array::from(Vec::<i32>::new()))])
                .unwrap();
        let target: SchemaRef = Arc::new(Schema::new(vec![Field::new("i", DataType::Int64, true)]));
        let batch = conform(empty, &target).unwrap();
        assert_eq!(batch.num_rows(), 0);
        assert_eq!(batch.schema(), target);
    }

    #[test]
    fn a_declared_schema_reaches_a_vec_of_batches() {
        let reader = vec![divergent(), divergent()]
            .into_batch_reader(None)
            .unwrap();
        let reader = conform_reader(reader, Some(declared())).unwrap();
        assert_eq!(reader.schema(), declared());
        assert_eq!(count(reader), 4);
    }
}
